lib/src/config.rs

```diff
@@ -381,10 +407,108 @@ impl StackedConfig {
     }
+
+    /// Looks up value of the specified type `T` from all layers, merges sub
+    /// fields as needed.
+    pub fn get<'de, T: Deserialize<'de>>(
+        &self,
+        name: impl ToConfigNamePath,
+    ) -> Result<T, ConfigError> {
+        self.get_item_with(name, T::deserialize)
+    }
+
+    /// Looks up value from all layers, merges sub fields as needed.
+    pub fn get_value(&self, name: impl ToConfigNamePath) -> Result<ConfigValue, ConfigError> {
+        self.get_item_with(name, Ok)
+    }
+
+    /// Looks up sub table from all layers, merges fields as needed.
+    // TODO: redesign this to attach better error indication?
+    pub fn get_table(&self, name: impl ToConfigNamePath) -> Result<ConfigTable, ConfigError> {
+        self.get(name)
+    }
+
+    fn get_item_with<T>(
+        &self,
+        name: impl ToConfigNamePath,
+        convert: impl FnOnce(ConfigValue) -> Result<T, ConfigError>,
+    ) -> Result<T, ConfigError> {
+        let name = name.into_name_path();
+        let name = name.borrow();
+        let (item, _layer_index) = get_merged_item(&self.layers, name)
+            .ok_or_else(|| ConfigError::NotFound(name.to_string()))?;
+        // TODO: Add source type/path to the error message. If the value is
+        // a table, the error might come from lower layers. We cannot report
+        // precise source information in that case. However, toml_edit captures
+        // dotted keys in the error object. If the keys field were public, we
+        // can look up the source information. This is probably simpler than
+        // reimplementing Deserializer.
+        convert(item).map_err(|err| err.extend_with_key(&name.to_string()))
+    }
+}
+
+/// Looks up item from `layers`, merges sub fields as needed. Returns a merged
+/// item and the uppermost layer index where the item was found.
+fn get_merged_item(
+    layers: &[ConfigLayer],
+    name: &ConfigNamePathBuf,
+) -> Option<(ConfigValue, usize)> {
+    let mut to_merge = Vec::new();
+    for (index, layer) in layers.iter().enumerate().rev() {
+        let item = match layer.lookup_item(name) {
+            Ok(Some(item)) => item,
+            Ok(None) => continue, // parent is a table, but no value found
+            Err(_) => break,      // parent is not a table, shadows lower layers
+        };
+        if matches!(item.kind, config::ValueKind::Table(_)) {
+            to_merge.push((item, index));
+        } else if to_merge.is_empty() {
+            return Some((item.clone(), index)); // no need to allocate vec
+        } else {
+            break; // shadows lower layers
+        }
+    }
+
+    // Simply merge tables from the bottom layer. Upper items should override
+    // the lower items (including their children) no matter if the upper items
+    // are shadowed by the other upper items.
+    let (mut merged, mut top_index) = match to_merge.pop() {
+        Some((item, index)) => (item.clone(), index),
+        None => return None,
+    };
```

Comment:
optional: I suppose this could be written using `?` for `Option`:

```rust
    let (item, mut top_index) = to_merge.pop()?;
    let mut merged = item.clone();
```

Reply:
Good catch. Maybe it was implemented that way because the return type was originally `Result<_>`.